File: samples/java/SupportChat/Server/Support/src/main/java/systems/zlink/samples/supportchat/server/support/handlers/AllocateConversationHandler.java
package systems.zlink.samples.supportchat.server.support.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkRequestHandler;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.samples.supportchat.server.configuration.SampleNames;
import systems.zlink.samples.supportchat.server.support.application.ConversationAllocator;
import systems.zlink.samples.supportchat.shared.contracts.Messages;

import java.util.concurrent.CompletionStage;

@ZLinkHandlerGroup(SampleNames.SupportChannel)
public final class AllocateConversationHandler
        implements ZLinkRequestHandler<
                Messages.AllocateConversationReq, Messages.AllocateConversationRes> {
    private final ConversationAllocator allocator;

    public AllocateConversationHandler(ConversationAllocator allocator) {
        this.allocator = allocator;
    }

    @Override
    public CompletionStage<Messages.AllocateConversationRes> handle(
            Messages.AllocateConversationReq request, ZLinkMessageContext context) {
        return allocator
                .allocate(
                        request.customerActorId(), request.customerDisplayName(), request.subject())
                .thenApply(
                        conversationId ->
                                new Messages.AllocateConversationRes(
                                        conversationId, SampleNames.Statuses.WaitingForAgent));
    }
}

File: samples/kotlin/SupportChat/Server/Support/src/main/kotlin/systems/zlink/samples/kotlin/supportchat/server/support/infrastructure/zlink/handlers/AllocateConversationHandler.kt
package systems.zlink.samples.kotlin.supportchat.server.support.infrastructure.zlink.handlers

import org.slf4j.LoggerFactory
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.handlers.ZLinkHandlerGroup
import systems.zlink.framework.kotlin.ZLinkSuspendingRequestHandler
import systems.zlink.samples.kotlin.supportchat.server.configuration.ConversationStatuses
import systems.zlink.samples.kotlin.supportchat.server.configuration.SampleNames
import systems.zlink.samples.kotlin.supportchat.server.support.application.SupportConversationAllocator
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AllocateConversationReq
import systems.zlink.samples.kotlin.supportchat.shared.contracts.AllocateConversationRes

@ZLinkHandlerGroup(SampleNames.SupportChannel)
class AllocateConversationHandler(private val allocator: SupportConversationAllocator) :
    ZLinkSuspendingRequestHandler<AllocateConversationReq, AllocateConversationRes> {
    override suspend fun handle(
        request: AllocateConversationReq,
        context: ZLinkMessageContext,
    ): AllocateConversationRes {
        logger.info(
            "support allocate: request customer={} subject={}",
            request.customerActorId,
            request.subject,
        )
        val conversationId =
            allocator.allocate(
                request.customerActorId,
                request.customerDisplayName,
                request.subject,
            )
        logger.info("support allocate: created conversation={}", conversationId)
        return AllocateConversationRes(conversationId, ConversationStatuses.WaitingForAgent)
    }

    private companion object {
        private val logger = LoggerFactory.getLogger(AllocateConversationHandler::class.java)
    }
}
